crates/metadata-store/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+#![allow(dead_code)]
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use restate_types::Version;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+#[derive(Debug, thiserror::Error)]
+pub enum NetworkError {
+    #[error("unreachable")]
+    Unreachable,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReadError {
+    #[error(transparent)]
+    Network(NetworkError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum WriteError {
+    #[error("failed condition: {0}")]
+    FailedCondition(String),
+    #[error(transparent)]
+    Network(NetworkError),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VersionedValue {
+    pub version: Version,
+    pub value: Bytes,
+}
+
+impl VersionedValue {
+    pub fn new(version: Version, value: Bytes) -> Self {
+        Self { version, value }
+    }
+}
+
+/// Conditions for the write operations of the [`MetadataStore`].
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Condition {
+    /// Key-value pair must not exist for the write operation to succeed.
+    DoesNotExist,
+    /// Key-value pair must have the provided [`Version`] for the write operation to succeed.
+    MatchesVersion(Version),
+}
+
+/// Metadata store abstraction. The metadata store implementations need to support linearizable
+/// reads and atomic compare and swap operations.
+#[async_trait]
+pub trait MetadataStore: Send + Sync {
```

Comment:
We might also need a `set()` that always overwrites (accepts a key, value, version) for admin operations.

Reply:
I was assuming that we'd want to keep them in sync, no? What's the motivation for maintaining two versions?

Reply:
I think it would only make sense to have two versions if we wanted to store values that don't have an inherent version. At the moment, I think all our metadata is versioned so it does not make sense for the metadata. Will the cluster controller store state that is not versioned (like the latest plan how to split and move partitions)? If this is not needed, then I could change the trait to take a `Versioned + Serialize` value.

Reply:
One thing I haven't checked is whether we could then still implement the `MetadataStore` against DynamoDB or some other key-value store that supports conditional write operations.

Reply:
I'm assuming that everything will be versioned, but even if not, wouldn't a `set()` API work for those (assuming that they'd always use a fixed version)?

Reply:
Also, does the trait itself need to be Send+Sync?

Reply:
> Also, does the trait itself need to be Send+Sync?

You are right, it is not needed.

> I'm assuming that everything will be versioned, but even if not, wouldn't a `set()` API work for those (assuming that they'd always use a fixed version)?

If the value would use a fixed version then it will become harder to tell what is the valid value if there are two writers (e.g. 2 cluster controllers that try to do contradicting operations). But I guess the answer would then be to add an explicit version to all state that is stored in the `MetadataStore`.